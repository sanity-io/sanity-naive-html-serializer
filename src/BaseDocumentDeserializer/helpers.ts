import {htmlToBlocks} from '@sanity/block-tools'
import {Schema} from '@sanity/schema'
import {ObjectField, PortableTextSpan, PortableTextTextBlock} from 'sanity'

const defaultSchema = Schema.compile({
  name: 'default',
  types: [
    {
      type: 'object',
      name: 'default',
      fields: [
        {
          name: 'block',
          type: 'array',
          of: [{type: 'block'}],
        },
      ],
    },
  ],
})

export const blockContentType = defaultSchema
  .get('default')
  .fields.find((field: ObjectField) => field.name === 'block').type

export const noSchemaWarning = (obj: Element): string =>
  `WARNING: Unfortunately the deserializer may have issues with this field or object: ${obj.className}.
  If it's a specific type, you may need to declare  at the top level, or write a custom deserializer.`

//helper to handle messy input -- take advantage
//of blockTools' sanitizing behavior for single strings
export const preprocess = (html: string): string => {
  const intermediateBlocks = htmlToBlocks(
    `<p>${html}</p>`,
    blockContentType
  ) as PortableTextTextBlock<PortableTextSpan>[]
  if (!intermediateBlocks.length) {
    throw new Error(`Error parsing string '${html}'`)
  }
  return intermediateBlocks[0].children[0].text
}
