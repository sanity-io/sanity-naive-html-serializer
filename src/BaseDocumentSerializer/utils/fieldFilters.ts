import {ObjectField, TypedObject} from 'sanity'

const META_FIELDS = ['_key', '_type', '_id']

/*
 * Helper. If field-level translation pattern used, only sends over
 * content from the base language. Works recursively, so if users
 * use this pattern several layers deep, base language fields will still be found.
 */
export const languageObjectFieldFilter = (
  obj: Record<string, any>,
  baseLang: string
): Record<string, any> => {
  const filterToLangField = (childObj: Record<string, any>) => {
    const filteredObj: Record<string, any> = {}
    filteredObj[baseLang] = childObj[baseLang]
    META_FIELDS.forEach((field) => {
      if (childObj[field]) {
        filteredObj[field] = childObj[field]
      }
    })
    return filteredObj
  }

  const findBaseLang = (childObj: Record<string, any>): Record<string, any> => {
    const filteredObj: Record<string, any> = {}
    META_FIELDS.forEach((field) => {
      if (childObj[field]) {
        filteredObj[field] = childObj[field]
      }
    })

    for (const key in childObj) {
      if (childObj.hasOwnProperty(key)) {
        const value: any = childObj[key]
        //we've reached a base language field, add it to
        //what we want to send to translation
        if (value.hasOwnProperty(baseLang)) {
          filteredObj[key] = filterToLangField(value)
        }
        //we have an array that may have language fields in its objects
        else if (Array.isArray(value) && value.length && typeof value[0] === 'object') {
          //recursively find and filter for any objects that have the base language
          const validLangObjects = value.reduce((validArr, objInArray) => {
            if (objInArray._type === 'block') {
              validArr.push(objInArray)
            } else if (objInArray.hasOwnProperty(baseLang)) {
              validArr.push(filterToLangField(objInArray))
            } else {
              const filtered = findBaseLang(objInArray)
              const nonMetaFields = Object.keys(filtered).filter(
                (objInArrayKey) => META_FIELDS.indexOf(objInArrayKey) === -1
              )
              if (nonMetaFields.length) {
                validArr.push(filtered)
              }
            }
            return validArr
          }, [])
          if (validLangObjects.length) {
            filteredObj[key] = validLangObjects
          }
        }
        //we have an object nested in an object
        //recurse down the tree
        else if (typeof value === 'object') {
          const nestedLangObj = findBaseLang(value)
          const nonMetaFields = Object.keys(nestedLangObj).filter(
            (nestedObjKey) => META_FIELDS.indexOf(nestedObjKey) === -1
          )
          if (nonMetaFields.length) {
            filteredObj[key] = nestedLangObj
          }
        }
      }
    }
    return filteredObj
  }

  //send top level object into recursive function
  return findBaseLang(obj)
}

/*
 * Eliminates stop-types and non-localizable fields
 * for document-level translation.
 */
export const fieldFilter = (
  obj: Record<string, any>,
  objFields: ObjectField[],
  stopTypes: string[]
): TypedObject => {
  const filteredObj: TypedObject = {_type: obj._type}

  const fieldFilterFunc = (field: Record<string, any>) => {
    if (field.localize === false) {
      return false
    } else if (field.type === 'string' || field.type === 'text') {
      return true
    } else if (Array.isArray(obj[field.name])) {
      return true
    } else if (!stopTypes.includes(field.type)) {
      return true
    }
    return false
  }

  const validFields = [
    ...META_FIELDS,
    ...objFields?.filter(fieldFilterFunc)?.map((field) => field.name),
  ]
  validFields.forEach((field) => {
    if (obj[field]) {
      filteredObj[field] = obj[field]
    }
  })
  return filteredObj
}
