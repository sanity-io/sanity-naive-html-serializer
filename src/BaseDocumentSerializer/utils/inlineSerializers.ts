import {PortableTextComponents} from '@portabletext/to-html'
import {ArbitraryTypedObject} from '@portabletext/types'
import {PortableTextSpan, PortableTextTextBlock, Schema} from 'sanity'
import {BaseDocumentSerializer} from '..'

type ObjectSerializerMethod = ({value}: {value: ArbitraryTypedObject}) => string
type MarkSerializerMethod = ({
  value,
  markType,
  children,
}: {
  value: ArbitraryTypedObject
  markType: string
  children: any[]
}) => string

const validateInlineContent = (
  obj: ArbitraryTypedObject | PortableTextSpan,
  serializers: PortableTextComponents,
  stopTypes: string[]
) =>
  //don't serialize blocks or spans
  obj._type !== 'block' &&
  obj._type !== 'span' &&
  //don't serialize any types that have been explicitly marked as stop types
  !stopTypes.includes(obj._type) &&
  //don't override custom serialization methods from the user
  !serializers.types?.[obj._type]

export const createInlineSerializerTypes = (
  block: PortableTextTextBlock,
  stopTypes: string[],
  serializers: PortableTextComponents,
  schemas: Schema
): PortableTextComponents => {
  const inlineObjects = block.children.filter((child) =>
    validateInlineContent(child, serializers, stopTypes)
  )
  const types: string[] = Array.from(new Set(inlineObjects.map((obj) => obj._type)))
  const serializerFunc: ObjectSerializerMethod = ({value}: {value: ArbitraryTypedObject}) => {
    const wrapperStart = [
      `<span class="default-serialized-object"`,
      `id="${value._key ?? value._id}"`,
      `data-type="${value._type}">`,
    ].join(' ')
    const serializedData = BaseDocumentSerializer(schemas).serializeObject(
      value,
      stopTypes,
      serializers
    )
    const wrapperEnd = `</span>`
    return [wrapperStart, serializedData, wrapperEnd].join('')
  }
  return types.reduce<Record<string, ObjectSerializerMethod>>((acc, type) => {
    acc[type as string] = serializerFunc
    return acc
  }, {})
}

export const createMarkSerializerTypes = (
  block: PortableTextTextBlock,
  stopTypes: string[],
  serializers: PortableTextComponents,
  schemas: Schema
): PortableTextComponents => {
  const marks = block.markDefs?.filter((mark) =>
    validateInlineContent(mark, serializers, stopTypes)
  )
  const markTypes = Array.from(new Set(marks?.map((mark) => mark._type)))
  const serializerFunc = ({
    value,
    markType,
    children,
  }: {
    value: ArbitraryTypedObject
    markType: string
    children: any[]
  }) => {
    const wrapperStart = [
      `<span class="default-serialized-annotation"`,
      `id="${value._key ?? value._id}"`,
      `data-type="${markType}">`,
    ].join(' ')
    const serializedChildren = `<span class="default-serialized-children">${children}</span>`
    const serializedData = BaseDocumentSerializer(schemas).serializeObject(
      value,
      stopTypes,
      serializers
    )
    const wrapperEnd = `</span>`
    return [wrapperStart, serializedChildren, serializedData, wrapperEnd].join('')
  }
  return markTypes.reduce<Record<string, MarkSerializerMethod>>((acc, type) => {
    acc[type as string] = serializerFunc
    return acc
  }, {})
}
