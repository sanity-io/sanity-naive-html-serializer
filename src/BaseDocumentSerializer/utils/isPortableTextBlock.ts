import {ArbitraryTypedObject} from '@portabletext/types'
import {PortableTextTextBlock} from 'sanity'

export const isPortableTextBlock = (el: ArbitraryTypedObject): el is PortableTextTextBlock => {
  return el._type === 'block'
}
