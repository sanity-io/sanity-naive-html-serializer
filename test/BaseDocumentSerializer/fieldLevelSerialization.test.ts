import {PortableTextBlock} from 'sanity'
import {getSerialized, getValidFields, toPlainText} from '../helpers'
import {fieldLevelArticle, findByClass, getHTMLNode, nestedLanguageFields} from './utils'

const serialized = getSerialized(fieldLevelArticle, 'field')
const docTree = getHTMLNode(serialized).body.children[0]

test('Global test of working field-level functionality and snapshot match', () => {
  expect(serialized).toMatchSnapshot()
})

test('String and text types get serialized correctly at top-level -- field level', () => {
  const titleObj = findByClass(docTree.children, 'title')?.children[0]
  const HTMLString = findByClass(titleObj!.children, 'en')
  const snippetObj = findByClass(docTree.children, 'snippet')?.children[0]
  const HTMLText = findByClass(snippetObj!.children, 'en')
  expect(HTMLString?.innerHTML).toEqual(fieldLevelArticle.title.en)
  expect(HTMLText?.innerHTML).toEqual(fieldLevelArticle.snippet.en)
})

describe('Presence and accuracy of fields in "vanilla" deserialization -- objects', () => {
  const getFieldLevelObjectField = () => {
    const config = findByClass(docTree.children, 'config')?.children[0]
    //return english field
    const englishConfig = findByClass(config!.children, 'en')
    return findByClass(englishConfig!.children, 'objectField')
  }

  const objectField = getFieldLevelObjectField()

  test('Top-level nested objects contain all serializable fields -- field level', () => {
    const fieldNames = getValidFields(fieldLevelArticle.config.en)
    const foundFieldNames = Array.from(objectField!.children).map((child) => child.className)

    expect(foundFieldNames.sort()).toEqual(fieldNames.sort())
  })

  test('Nested object in object contains all serializable fields -- field Level', () => {
    const nestedObject = findByClass(objectField!.children, 'objectAsField')!.children[0]
    const fieldNames = getValidFields(fieldLevelArticle.config.en.objectAsField)
    const foundFieldNames = Array.from(nestedObject!.children).map((child) => child.className)
    expect(foundFieldNames.sort()).toEqual(fieldNames.sort())
  })

  test('Nested object contains accurate values -- field level', () => {
    const title = fieldLevelArticle.config.en.title
    const blockText = toPlainText(fieldLevelArticle.config.en.nestedArrayField)

    expect(objectField?.innerHTML).toContain(title)
    expect(objectField?.innerHTML).toContain(blockText)
  })

  test('Nested object in an object contains accurate values -- field level', () => {
    const nestedObject = findByClass(objectField!.children, 'objectAsField')!.children[0]
    const title = fieldLevelArticle.config.en.objectAsField.title
    const blockText = toPlainText(fieldLevelArticle.config.en.objectAsField.content)

    expect(nestedObject.innerHTML).toContain(title)
    expect(nestedObject.innerHTML).toContain(blockText)
  })
})

/*
 * Presence and accuracy of fields in "vanilla" deserialization -- arrays
 */
describe('Presence and accurancy of fields in "vanilla" deserialization -- arrays', () => {
  const getFieldLevelArrayField = () => {
    const content = findByClass(docTree.children, 'content')?.children[0]
    return findByClass(content!.children, 'en')
  }
  const arrayField = getFieldLevelArrayField()

  test('Array contains all serializable blocks with keys, in order -- field level', () => {
    const origKeys = fieldLevelArticle.content.en.map((block: PortableTextBlock) => block._key)
    const serializedKeys = Array.from(arrayField!.children).map((block) => block.id)
    expect(serializedKeys).toEqual(origKeys)
  })

  test('Array contains top-level block text -- field level', () => {
    const blockText = toPlainText(fieldLevelArticle.content.en).trim()
    expect(arrayField?.innerHTML).toContain(blockText)
  })

  test('Object in array contains all serializable fields -- field level', () => {
    const objectInArray = findByClass(arrayField!.children, 'objectField')
    const fieldNames = getValidFields(
      fieldLevelArticle.content.en.find(
        (block: Record<string, any>) => block._type === 'objectField'
      )
    )
    const foundFieldNames = Array.from(objectInArray!.children).map((child) => child.className)
    expect(foundFieldNames.sort()).toEqual(fieldNames.sort())
  })

  test('Object in array contains accurate values in nested object -- field level', () => {
    const objectInArray = findByClass(arrayField!.children, 'objectField')
    const nestedObject = findByClass(objectInArray!.children, 'objectAsField')
    const title = fieldLevelArticle.content.en.find(
      (block: Record<string, any>) => block._type === 'objectField'
    ).objectAsField.title
    const blockText = toPlainText(
      fieldLevelArticle.content.en.find(
        (block: Record<string, any>) => block._type === 'objectField'
      ).objectAsField.content
    ).trim()
    expect(nestedObject?.innerHTML).toContain(title)
    expect(nestedObject?.innerHTML).toContain(blockText)
  })
})

test('Nested locale fields make it to serialization, but only base lang', () => {
  const nestedLocales = {...fieldLevelArticle, ...nestedLanguageFields}
  const nestedSerialized = getSerialized(nestedLocales, 'field')
  const nestedDocTree = getHTMLNode(nestedSerialized).body.children[0]
  const slices = findByClass(nestedDocTree.children, 'slices')
  const pageFields = findByClass(nestedDocTree.children, 'pageFields')
  expect(slices?.innerHTML).toContain(nestedLanguageFields.slices[0].en[0].children[0].text)
  expect(pageFields?.innerHTML).toContain(nestedLanguageFields.pageFields.name.en)
  expect(slices?.innerHTML).not.toContain(nestedLanguageFields.slices[0].fr_FR[0].children[0].text)
  expect(pageFields?.innerHTML).not.toContain(nestedLanguageFields.pageFields.name.fr_FR)
})

//also test: setting different base language!
