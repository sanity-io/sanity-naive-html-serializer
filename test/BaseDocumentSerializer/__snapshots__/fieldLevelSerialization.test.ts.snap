// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global test of working field-level functionality and snapshot match 1`] = `
Object {
  "content": "<html><head><meta name=\\"_id\\" content=\\"drafts.2947533e-1ea5-4116-955b-339608d3445d\\"><meta name=\\"_type\\" content=\\"fieldLevelArticle\\"><meta name=\\"_rev\\" content=\\"l58oha-n06-1s4-f6i-74g6cuakl\\"><meta name=\\"version\\" content=\\"3\\"></head><body><div class=\\"fieldLevelArticle\\" id=\\"drafts.2947533e-1ea5-4116-955b-339608d3445d\\" data-type=\\"object\\"><div class=\\"title\\" data-level=\\"field\\"><div class=\\"\\" data-type=\\"object\\"><span class=\\"en\\">My Field-Level Article</span></div></div><div class=\\"snippet\\" data-level=\\"field\\"><div class=\\"\\" data-type=\\"object\\"><span class=\\"en\\">This is text in my text field</span></div></div><div class=\\"tags\\" data-level=\\"field\\"><div class=\\"\\" data-type=\\"object\\"><div class=\\"en\\" data-type=\\"array\\"><span>tag 1</span><span>tag 2</span><span>tag 3</span></div></div></div><div class=\\"config\\" data-level=\\"field\\"><div class=\\"\\" data-type=\\"object\\"><div class=\\"en\\" data-level=\\"field\\"><div class=\\"objectField\\" data-type=\\"object\\"><span class=\\"title\\">This is an object nested in a document</span><div class=\\"objectAsField\\" data-level=\\"field\\"><div class=\\"childObjectField\\" data-type=\\"object\\"><span class=\\"title\\">This is one level deeper</span><div class=\\"content\\" data-type=\\"array\\"><p id=\\"869168fe3a2a\\">This is block text 2 levels deep</p></div></div></div><div class=\\"nestedArrayField\\" data-type=\\"array\\"><p id=\\"f49b4d7e3e51\\">This is block text 1 level deep</p></div></div></div></div></div><div class=\\"content\\" data-level=\\"field\\"><div class=\\"\\" data-type=\\"object\\"><div class=\\"en\\" data-type=\\"array\\"><p id=\\"e2a39d768ff8\\">This is block text at the top level.</p><div class=\\"objectField\\" id=\\"271b0c6ee984\\" data-type=\\"object\\"><span class=\\"title\\">This is an object in top-level block text.</span><div class=\\"objectAsField\\" data-level=\\"field\\"><div class=\\"childObjectField\\" data-type=\\"object\\"><span class=\\"title\\">This is a nested object in an object in top-level block text.</span><div class=\\"content\\" data-type=\\"array\\"><p id=\\"4f24f6fbfae7\\">This is block text in a nested object in an object in top-level block text.</p></div></div></div></div></div></div></div></div></body></html>",
  "name": "drafts.2947533e-1ea5-4116-955b-339608d3445d",
}
`;
